### DISCUSSION TITLE
I can get an MQTT SSL connection,  or a regular connection via Proxy, but not both (SSL via Proxy)

### DISCUSSION THREAD

**Discussion by oreillymj** [netty/netty#14224] [Q&A]
So I have been testing connections to [this ](https://www.emqx.com/en/mqtt/public-mqtt5-broker) public MQTT broker using Netty as a client.

I can connect on the insecure 1883 port with and without a SOCK5 proxy in my pipeline.

I can also connect to the secure port 8883 with the ca cert [here](https://assets.emqx.com/data/broker.emqx.io-ca.crt) after pulling down the full chain using this [web-site](https://whatsmychaincert.com/?broker.emqx.io:8883).

But what I cannot do is get the SSL connection working via the SOCKS proxy. Other MQTT clients like mosquitto_sub work via the proxy, so I do know that it should in theory work. e.g. (proxy address obscured)

```
mosquitto_sub -v -d -I mosq_sub_client -h broker.emqx.io -p 8883 -t test/topic --proxy socks5h://none@proxy-server.example.com:1080 --cafile c:/temp/broker.emqx.io-ca.crt
```
With Netty 4.1.111, I get this error....
```
2024-08-12 15:55:37:137 - MqttPingHandler->userEventTriggered()-> SslHandshakeCompletionEvent(java.nio.channels.ClosedChannelException)
MqttHandler()->exceptionCaught->socks5, none, proxy-server.example.com.com/XXX.XXX.XXX.XXX:1080 => broker.emqx.io/54.244.173.190:8883, java.nio.channels.ClosedChannelException
MqttHandler()->userEventTriggered()->
MqttHandler()->userEventTriggered()-> SslCloseCompletionEvent(java.nio.channels.ClosedChannelException)
2024-08-12 15:55:37:145 - MqttPingHandler->userEventTriggered()->
2024-08-12 15:55:37:145 - MqttPingHandler->userEventTriggered()-> SslCloseCompletionEvent(java.nio.channels.ClosedChannelException)
io.netty.handler.proxy.ProxyConnectException: socks5, none, proxy-server.example.com.com/XXX.XXX.XXX.XXX:1080 => broker.emqx.io/54.244.173.190:8883, java.nio.channels.ClosedChannelException
.........

Caused by: java.nio.channels.ClosedChannelException
	at io.netty.handler.ssl.SslHandler.channelInactive(SslHandler.java:1154)
	... 16 more
	Suppressed: io.netty.handler.ssl.StacklessSSLHandshakeException: Connection closed while SSL/TLS handshake was in progress
		at io.netty.handler.ssl.SslHandler.channelInactive(Unknown Source)
```

Timeouts are set to 60 seconds, but this error occurs much quicker.
Is there something I need to do to tell the SSL handler to wait for the proxy connection to be established before attempting to perform the SSL handshake?

**Comment by oreillymj** [comment]
TLDR; Double check that the Proxy is the first handler in the pipeline. In my case SSL with addFirst jumped ahead of Proxy which was not the intent.

ok. both SSL and Proxy handlers were optional handlers being added to the pipeline. And proxy was being added to the pipeline before SSL so I thought it was First in the pipeline, but SSL was also being added with .addFirst()
But with Proxy and then SSL both being added with addFirst(), SSL effectively is the start of the pipeline instead of Proxy.
This [post](https://stackoverflow.com/questions/35119032/how-to-use-socks4-5-proxy-handlers-in-netty-client-4-1) helped me realise the mistake in the pipeline setup.

**Comment by oreillymj** [comment]
mqtt://broker.emqx.io:1883 - works with and without a proxy

ssl://broker.emqx.io:8883 - SSL works with and without a proxy

ws://broker.emqx.io:8083/mqtt works with danted socks5 proxy.

wss://broker.emqx.io:8084 - SSL works with and without a proxy
